autoblocks/_impl/testing/run.py

```diff
@@ -0,0 +1,301 @@
+        return
+
+    await client.post(
+        "/results",
+        json=dict(
+            testExternalId=test_id,
+            testCaseHash=test_case._cached_hash,
+            testCaseBody=dataclasses.asdict(test_case),
+            testCaseOutput=output,
+        ),
+    )
+
+    try:
+        await gather_with_max_concurrency(
+            max_evaluator_concurrency,
+            *[
+                evaluate_output(
+                    test_id=test_id,
+                    test_case=test_case,
+                    output=output,
+                    evaluator=evaluator,
+                )
+                for evaluator in evaluators
+            ],
+        )
+    except Exception as err:
+        await send_error(
+            test_id=test_id,
+            test_case_hash=test_case._cached_hash,
+            evaluator_id=None,
+            error=err,
+        )
+
+
+async def async_run_test_suite(
+    test_id: str,
+    test_cases: List[BaseTestCase],
+    evaluators: List[BaseEvaluator],
+    fn: Callable,
+    max_test_case_concurrency: int,
+    max_evaluator_concurrency: int,
+):
+    try:
+        assert test_cases, f"[{test_id}] No test cases provided."
+        for test_case in test_cases:
+            assert isinstance(
+                test_case,
+                BaseTestCase,
+            ), f"[{test_id}] Test case {test_case} does not implement BaseTestCase."
+        for evaluator in evaluators:
+            assert isinstance(
+                evaluator,
+                BaseEvaluator,
+            ), f"[{test_id}] Evaluator {evaluator} does not implement BaseEvaluator."
+    except Exception as err:
+        await send_error(
+            test_id=test_id,
+            test_case_hash=None,
+            evaluator_id=None,
+            error=err,
+        )
+        return
+
+    await client.post("/start", json=dict(testExternalId=test_id))
+
+    try:
+        await gather_with_max_concurrency(
+            max_test_case_concurrency,
+            *[
+                run_test_case(
+                    test_id=test_id,
+                    test_case=test_case,
+                    evaluators=evaluators,
+                    fn=fn,
+                    max_evaluator_concurrency=max_evaluator_concurrency,
+                )
+                for test_case in test_cases
+            ],
+        )
+    except Exception as err:
+        await send_error(
+            test_id=test_id,
+            test_case_hash=None,
+            evaluator_id=None,
+            error=err,
+        )
+
+    await client.post("/end", json=dict(testExternalId=test_id))
+
+
+def run_test_suite(
+    id: str,
+    test_cases: List[BaseTestCase],
+    evaluators: List[BaseEvaluator],
+    fn: Callable,
+    # How many test cases to run concurrently
+    max_test_case_concurrency: int = 10,
+    # How many evaluators to run concurrently on the result of a test case
+    max_evaluator_concurrency: int = 5,
+):
+    init_global_vars()
```

Comment:
nit: I think this function name should also mention the fact that it starts the event loop